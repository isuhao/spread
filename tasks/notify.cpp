#include "notify.hpp"
#include <assert.h>

using namespace Tasks;
using namespace Jobify;

NotifyTask::NotifyTask(Jobify::Job *j)
  : other(j)
{
  assert(other);

  // This discards our own info struct, but that's OK.
  info = other->getInfo();
}

void NotifyTask::doJob()
{
  // Make sure we reset the info status before running the other
  // job, or it will complain that it has already started.
  assert(info->status == ST_BUSY);
  info->status = ST_CREATED;

  other->run();
  assert(!info->isBusy());
}

void NotifyTask::cleanup()
{
  if(info->isSuccess()) onSuccess();
  else onError();
}
