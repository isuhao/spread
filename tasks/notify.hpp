#ifndef __TASKS_NOTIFY_HPP_
#define __TASKS_NOTIFY_HPP_

#include <job/job.hpp>

namespace Tasks
{
  /* NotifyTask is meant to be subclassed, and is used for getting
     notifications when another task succeeds or fails.

     Pass along a job to the constructor, and run it as normal. The
     NotifyTask will share the other job's JobInfo.

     The destructor deletes the job.
   */
  struct NotifyTask : Jobify::Job
  {
    NotifyTask(Jobify::Job *j);
    ~NotifyTask() { delete other; }

  protected:
    // One of these are called when the job has finished
    virtual void onSuccess() {}
    virtual void onError() {}

  private:
    Jobify::Job *other;

    void doJob();
    void cleanup();
  };
}

#endif
