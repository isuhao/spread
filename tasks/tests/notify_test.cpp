#include "notify.hpp"

#include <iostream>
#include <stdexcept>

using namespace std;
using namespace Tasks;
using namespace Jobify;

struct MyJob : Job
{
  int i;

  MyJob(int _i) : Job(), i(_i) {}

  void doJob()
  {
    cout << "Job " << i << ": ";
    if(i == 1)
      {
        cout << "OK (normal execution)\n";
        setDone();
      }
    else if(i == 2)
      {
        cout << "ERROR! (using setError)\n";
        setError("We can't have that crap!");
      }
    else
      {
        cout << "EXCEPTION! (throw)\n";
        throw runtime_error("blablah");
      }
  }

  void cleanup()
  {
    cout << "Cleaning up " << i;
    if(info->isError()) cout << " (Error: " << info->message << ")\n";
    else cout << " (Success!)\n";
  }
};

struct TestNote : NotifyTask
{
  TestNote(int i)
    : NotifyTask(new MyJob(i))
  {}

  void onSuccess() { cout << "SUCCESS notified\n"; }
  void onError() { cout << "ERROR notified: " << info->message << "\n"; }
};

int main()
{
  TestNote a(1), b(2), c(3);

  a.run();
  b.run();
  c.run();

  return 0;
}
