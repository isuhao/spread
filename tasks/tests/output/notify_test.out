Job 1: OK (normal execution)
Cleaning up 1 (Success!)
SUCCESS notified
Job 2: ERROR! (using setError)
Cleaning up 2 (Error: We can't have that crap!)
ERROR notified: We can't have that crap!
Job 3: EXCEPTION! (throw)
Cleaning up 3 (Error: blablah)
ERROR notified: blablah
